### PR TITLE
add: example on training mlms from scratch on tpus.

### PR DESCRIPTION
**Pull request by sayakpaul**
Colab Notebook: 
https://colab.research.google.com/drive/1wHz14lg93dmGsMu_RZoUWFwdVxBWX9Xe?usp=sharing

Cc: @RocketKnight1

### CODE REVIEW

File: examples/nlp/mlm_training_tpus.py [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+"""
+Then, let's import the modules we need.
+"""
+
+import os
+import re
+
+import tensorflow as tf
+
+from transformers import (
```
Comment (fchollet) [inline examples/nlp/mlm_training_tpus.py:139]: Why not just `import transformers` and then access objects from there?

File: examples/nlp/mlm_training_tpus.py [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+# probability with which we mask the input tokens in a sequence.
+mlm_probability = 0.15
+data_collator = DataCollatorForLanguageModeling(
+    tokenizer=tokenizer, mlm_probability=mlm_probability, mlm=True, return_tensors="tf"
+)
+
+
+def mask_with_collator(batch):
+    special_tokens_mask = (
+        ~tf.cast(batch["attention_mask"], tf.bool)
```
Comment (fchollet) [inline examples/nlp/mlm_training_tpus.py:260]: You don't get to see `~` and `|` operators very often!

File: examples/nlp/mlm_training_tpus.py [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+)
+total_train_steps = steps_per_epoch * num_epochs
+learning_rate = 0.0001
+weight_decay_rate = 1e-3
+
+with strategy.scope():
+    model = TFAutoModelForMaskedLM.from_config(config)
+    model(
+        model.dummy_inputs
+    )  # Pass some dummy inputs through the model to ensure all the weights are built
```
Comment (fchollet) [inline examples/nlp/mlm_training_tpus.py:394]: This is not very elegant -- canonically you'd use the `build(input_shape)` method to build the model's state. In this case I suppose you could implement it as:

```
def build(self, input_shape=None):
    self(self.dummy_inputs)
    self.built = True
```

Then you can just do: `model.build()`.

File: examples/nlp/mlm_training_tpus.py [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+    optimizer, schedule = create_optimizer(
```
Comment (fchollet) [inline examples/nlp/mlm_training_tpus.py:395]: This is a bit blackboxy -- at least it would deserve a comment

File: examples/nlp/mlm_training_tpus.py [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+        num_train_steps=total_train_steps,
+        num_warmup_steps=total_train_steps // 20,
+        init_lr=learning_rate,
+        weight_decay_rate=weight_decay_rate,
+    )
+    model.compile(optimizer=optimizer, metrics=["accuracy"])
```
Comment (fchollet) [inline examples/nlp/mlm_training_tpus.py:401]: Worth explaining the lack of a loss argument here

**Review by fchollet — COMMENTED** [review]
Thank you for the PR! This is great stuff 👍

File: examples/nlp/mlm_training_tpus.py (hunk 3 above)
Comment (sayakpaul) [reply examples/nlp/mlm_training_tpus.py:387]: I will leave this for @RocketKnight1 to comment further.

**Comment by sayakpaul** [comment]
@fchollet thanks for your comments and I just addressed them!

Regarding https://github.com/keras-team/keras-io/pull/1381#discussion_r1199820427, I think it needs a separate design discussion which we'll keep in mind.

File: examples/nlp/mlm_training_tpus.py (hunk 3 above)
Comment (fchollet) [reply examples/nlp/mlm_training_tpus.py:387]: > I've been contemplating a community push to add proper build() methods everywhere, but every layer would need its own ones and it's quite a lot of code.

Why not just have one `build()` on the top-level model, implemented using your dummy inputs system?

> Also, do we need to set self.built after self(self.dummy_inputs)? I thought that build() would be called on the model and on each layer when the network is first called, which would set self.built for us.

Actually yes, it will already be set after `__call__()`.

Comment (Rocketknight1) [reply examples/nlp/mlm_training_tpus.py:387]: Ah, I didn't even think of that! I assumed it was somehow illegal to pass inputs through the model in the `build()` method! That's actually quite helpful for getting us more in-line with Keras best practices - I'll experiment with it and possibly file a PR soon.

Comment (fchollet) [reply examples/nlp/mlm_training_tpus.py:387]: >  I assumed it was somehow illegal to pass inputs through the model in the build() method! 

Ideally `build()` should be zero-flops -- so you can pass symbolic inputs, but hopefully not *eager* inputs.

You're allowed to do whatever you want in `build()`, there's just one rule: the only side effects should be building the model's variables. There should be no other side effect, ideally (e.g. no change to the values of existing non-trainable variables like BN stats, etc.).

**Review by fchollet — APPROVED** [review]
The PR looks good! You can add the generated files.

**Comment by sayakpaul** [comment]
@fchollet should be good for you to take the reigns here!

**Review by fchollet — APPROVED** [review]
Thank you for the contribution! 👍
